### ISSUE TITLE
How clangd works with multiple projects

### ISSUE THREAD

**Issue by ineffective** [clangd/clangd#317]
Hello, I have multiple projects (so multiple compilation_commands.json files). They are all quite small, but share enormous common code. Compilation commands files are basically identical when it comes to flags and toolchain - they have to, as CMakeLists.txt uses same toolchain file and generates nearly identical compilation_commands.json files. But it seems that everytime I open different project I have to wait for clangd to index everything - sometimes it takes more than hour. Does it sound like misconfiguration on my part or is it to be expected?
Would it be viable to have clangd _somehow_ run as a service that keeps everything in memory without need for restart everytime editor is closed?
And then would it be possible to have clangd run over the network? It would be great - we could setup central server with most of libraries (if not all) indexed and have clangd running there all the time - in such case it could even keep multiple indexes (for different configurations of libraries) and serve them depending on who wants what.
And since I'm dreaming, it would be great to have hierarchy of clangd servers, where central server has knowledge of all libraries and smaller servers (local, on client's machine) have knowledge of project-specific files only.

**Comment by kadircet** [comment]
> Hello, I have multiple projects (so multiple compilation_commands.json files). They are all quite small, but share enormous common code. Compilation commands files are basically identical when it comes to flags and toolchain - they have to, as CMakeLists.txt uses same toolchain file and generates nearly identical compilation_commands.json files. But it seems that everytime I open different project I have to wait for clangd to index everything - sometimes it takes more than hour. Does it sound like misconfiguration on my part or is it to be expected?
Would it be viable to have clangd somehow run as a service that keeps everything in memory without need for restart everytime editor is closed?

clangd should only perform that indexing once, and reuse it in the next runs(at least for the files that haven't changed). Could you provide more information about the directory layout and where compile_commands.json are located ?
In theory if the common code resides in a single directory and used by all other projects. that common code should only be indexed once.

> And then would it be possible to have clangd run over the network? It would be great - we could setup central server with most of libraries (if not all) indexed and have clangd running there all the time - in such case it could even keep multiple indexes (for different configurations of libraries) and serve them depending on who wants what.
And since I'm dreaming, it would be great to have hierarchy of clangd servers, where central server has knowledge of all libraries and smaller servers (local, on client's machine) have knowledge of project-specific files only.

we are planning to work on something like this soon @usx95 and @kirillbobyrev can provide more information about those.

**Comment by ineffective** [comment]
> > Hello, I have multiple projects (so multiple compilation_commands.json files). They are all quite small, but share enormous common code. Compilation commands files are basically identical when it comes to flags and toolchain - they have to, as CMakeLists.txt uses same toolchain file and generates nearly identical compilation_commands.json files. But it seems that everytime I open different project I have to wait for clangd to index everything - sometimes it takes more than hour. Does it sound like misconfiguration on my part or is it to be expected?
> > Would it be viable to have clangd somehow run as a service that keeps everything in memory without need for restart everytime editor is closed?
> 
> clangd should only perform that indexing once, and reuse it in the next runs(at least for the files that haven't changed). Could you provide more information about the directory layout and where compile_commands.json are located ?

compile_commands.json are located at the root of the project, this is where I start vim - I'm not sure what is the exact process of looking for these files and it seems that it always works. Never the less I have one gigantic directory where (cross-)compilers files reside (it is g++-9.2.0, so libstdc++) and other libraries (boost, libpqxx, protobuf, mqtt-cpp and a lot of others). When some library project reaches its target (so features are implemented and so on), it is packed into .deb file and installed - it is installed into this gigantic directory. 
Other projects, ones that produce binaries and no external libraries, do not use anything except for stuff from this directory (and system headers for proper architecture, which reside in e.g. /usr/local/include) and each has its own compiler_commands.json file generated by CMake.
If clangd indeed reuses its compilation database (index files?), then most probably it is classical example of PEBKAC - I might have misconfigured something. I used clangd-10 from Debian repos provided by the clangd project.

> > And then would it be possible to have clangd run over the network? It would be great - we could setup central server with most of libraries (if not all) indexed and have clangd running there all the time - in such case it could even keep multiple indexes (for different configurations of libraries) and serve them depending on who wants what.
> > And since I'm dreaming, it would be great to have hierarchy of clangd servers, where central server has knowledge of all libraries and smaller servers (local, on client's machine) have knowledge of project-specific files only.
> 
> we are planning to work on something like this soon @usx95 and @kirillbobyrev can provide more information about those.

That would be absolutely awesome, especially if few users could use same running service and "federation" was supported.

**Comment by kadircet** [comment]
Sorry I couldn't infer the information i was looking for from your description, let me put it this way:

## Case 1
Let's say you've got a project in "/repos/common/" with a comp_cmds at "/repos/common/compile_commands.json".

then you've got some projects in "/repos/proj1/" and "/repos/proj2/" and have respective comp_cmds.json files in their source root.

Let's say you've opened "/repos/proj1/a.cc" clangd will index any file mentioned in "/repos/proj1/compile_commands.json" and resides under "/repos/proj1" and produce cache files inside "/repos/proj1/.clangd/index/". So it won't(well, shouldn't) index anything inside "/repos/common/". Note that it will still produce some index files for headers inside the common project, since they are indexed as a result of files in proj1 depending on them, but it won't parse any source file inside common project, unless you open a file in that project.

Once you open a file inside common project, it will trigger the same process for /repos/common/compile_commands.json.

While indexing any source file mentioned in a compile_commands.json, clangd will first check the cached version of that file and skip indexing if file contents haven't changed.

## Case 2

"/repos/proj1" containing a subdirectory "/repos/proj1/common" and similarly for "/repos/proj2/" you've got a subdir "/repos/proj2/common" and there are comp_cmd.json files in:

"/repos/proj1/compile_commands.json"
"/repos/proj1/common/compile_commands.json"
"/repos/proj2/compile_commands.json"
"/repos/proj2/common/compile_commands.json"

once you open, /repos/proj1/a.cc clangd will again index everything inside proj1 directory except common subdirectory, as it has its own compile_commands.json. once you open a file inside "/repos/proj1/common/" it will be indexed similarly. but since "/repos/proj1/common" and "/repos/proj2/common" has their own compile_commands.json these projects won't be tied together, therefore won't share any cache hence will need to be indexed separately.

**Comment by ineffective** [comment]
> Sorry I couldn't infer the information i was looking for from your description

Sorry for that, I weren't sure what you were looking for and probably missed some (all :] ) details. 

What you describe as `Case 1` is how I have things set up. Let's say my setup is as follows:

/opt/common/
/repos/proj1
/repos/proj2
/repos/proj3

`/opt/common/` is a directory where all custom libraries and compilers reside - paths to includes are shared via toolchain file by all CMakeLists.txt files in all `/repos/projX` projects. Note that it contains only headers - these are installed libraries.

Now each `/repos/proj1` has its own `compile_commands.json` file generated by CMakeLists.txt. Basically any file in that projects `#include`s something from `/opt/common/` directory - and in most cases this includes `libstdc++` headers and `boost` headers. If my understanding is correct, clangd indexes file that I open (let's say `main.cpp`) and all its includes. This means that I have about half of directory `/opt/common/` indexed separately for each project? This would explain why it takes so long after opening each file - since these projects are really small (most of them have three, four source fiels) but there is a lot of them it is a lot of indexing, especially since standard library and boost are used everywhere - I'll check it again because now it seems to me that it may be just my perception.

To be sure: in my case above (three projects, one common directory), `/opt/common` will be indexed three times, once for every `/repos/projX` directory?

And I just realised what you wrote in `Case 2` - would it be possible to provide "fake" compile_commands.json file in root of `/opt/common` directory and somehow force clangd to index it? This way it would be indexed only once and most of my problems would go away. 

Do I have it right or am I _still_ missing something?

**Comment by kadircet** [comment]
**tl;dr;** The slowness you are facing seems to be resulting from parsing files, rather than background-indexing. And clangd doesn't implement any re-use of "parsed headers" while parsing a file, therefore it needs to happen at least once whenever you open a new file.


> To be sure: in my case above (three projects, one common directory), /opt/common will be indexed three times, once for every /repos/projX directory?

yes and no, there's no such a thing for "/opt/common/" to be indexed if it doesn't have any comp_cmds.json associated with it(or any source files).
but while indexing any source file in "/repos/projX" clangd needs to build the whole source file with all of its transitive includes (called a translation unit, TU for short) which apparently results in compiling half of the stuff inside /opt/common (looking at your description).

you can pass `-log=verbose` to the clangd and see which files are being indexed.

> And I just realised what you wrote in Case 2 - would it be possible to provide "fake" compile_commands.json file in root of /opt/common directory and somehow force clangd to index it? This way it would be indexed only once and most of my problems would go away.

as explained above, unfortunately not. there's no such a thing as re-using cache for headers while parsing a TU in clangd (as of today), and anything in /opt/common is parsed a side effect of being included by some source file inside /repos/projX/. in case of missing comp_cmds.json files, clangd puts the index into $HOME/.clangd/index.

> Do I have it right or am I still missing something?

looks correct to me, i hope i managed to provide some useful information :D

**Comment by kirillbobyrev** [comment]
> > > And then would it be possible to have clangd run over the network? It would be great - we could setup central server with most of libraries (if not all) indexed and have clangd running there all the time - in such case it could even keep multiple indexes (for different configurations of libraries) and serve them depending on who wants what.
> > > And since I'm dreaming, it would be great to have hierarchy of clangd servers, where central server has knowledge of all libraries and smaller servers (local, on client's machine) have knowledge of project-specific files only.
> > 
> > 
> > we are planning to work on something like this soon @usx95 and @kirillbobyrev can provide more information about those.
> 
> That would be absolutely awesome, especially if few users could use same running service and "federation" was supported.
Yes, I think that should be possible. We're working on this feature this quarter. We'll make an announcement and write the docs once it's possible!

**Comment by ineffective** [comment]
@kirillbobyrev Thanks, I somehow missed your answer. This would be absolutely awesome. I'll close this ticket and patiently wait for news regarding this feature.
